text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,639 @@
+requirements sketched above can be met in a consistent way across messaging
+scenarios and different messaging systems.
+
+## Terminology
+
+For terms used in this document, refer to [OTEP 173](#0173-messaging-semantic-conventions.md#terminology).
+
+## Stages of producing and consuming messages
+
+Producing and consuming a message involves five stages:
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  CR[Create] --> PU[Publish]
+  end
+  subgraph INTERMEDIARY
+  direction TB
+  end
+  subgraph CONSUMER
+  direction TB
+  RE[Receive] --> PR[Process]
+  PR --> SE[Settle]
+  end
+  PU --> INTERMEDIARY
+  INTERMEDIARY --> RE
+  SE -..-> INTERMEDIARY
+```
+
+1. The producer creates a message.
+2. The producer publishes the message to an intermediary.
+3. The consumer receives the message from an intermediary.
+4. The consumer processes the message.
+5. The consumer settles the message by notifying the intermediary that the
+   message was processed. In some cases (fire-and-forget scenarios, or when
+   settlement happens on the broker), the settlement stage does not exist.
+
+The semantic conventions described below define how to model those stages with
+spans.
+
+## Trace structure
+
+### Producer
+
+Producers are responsible for injecting a creation context into a message.
+Subsequent consumers will use this context to link consumer traces to producer
+traces. Ideally, each message gets a unique and distinct creation context
+assigned. However, as a context must refer to a span this would require the
+creation of a distinct span for each message, which is not feasible in all
+scenarios. In certain batching scenarios where many messages are created and
+published in large batches, creating a span for every single message would
+obfuscate traces and is not desirable. Thus instrumentation libraries and
+auto-instrumentation should default to creating a unique and distinct context
+per message, but may support configuration or other ways to change this default
+behavior.
+
+For each producer scenario, a "Publish" span needs to be created. This span
+measures the duration of the call or operation that provides messages for
+sending or publishing to an intermediary. This call or operation (and the
+related "Publish" span) can either refer to a single message or a batch of
+multiple messages.
+
+It is recommended to create a "Create" span for every single message. "Create"
+spans can be created during the "Publish" operation as children of the
+"Publish" span. Alternatively, "Create" spans can be created independently of
+the "Publish" operation. In this case, the "Publish" span should link to the
+"Create" spans.
+
+If a "Create" span exists for a message, its context must be injected into the
+message. If no "Create" span exists for a message, the context of the related
+"Publish" span must be injected into the message.
+
+"Create" spans must not be created for forwarded messages into which a context
```

Comment:
Can't we simply say here that Create spans must not be created/changed if the message already has one attached? Not sure "forwarded" is a concept that can be know during runtime? We simply want to enforce that it's immutable and once set, it can only be read, but not modified.

Reply:
Suggestion:

```
A "Create" context in a message is immutable. 
Once set, it should not change, thus a "Create" context should 
not be created for messages that already have one set.
```

Reply:
I'll remove the word "forwarded", I think you're right that it's out of place.

What I'd like to convey in this sentence, is that there can only be one "Create" span for each message. I'd like to avoid talking about replacing context here, I think we made that clear in the context propagation part that's already merged into the spec. I'd like to make clear that for one message, there should only be one "Create" _span_.

With your proposed wording, that wouldn't be clear: it would allow to have additional "Create" spans, as long as the context in the message is not touched.